src/Utility/Test/TestFunctions.C

```diff
@@ -46,6 +47,548 @@ TEST(TestScalarFunc, ParseDerivative)
+        for (size_t i = 1; i <= 2; ++i)
+          for (size_t j = 1; j <= 2; ++j)
+            EXPECT_DOUBLE_EQ(fx(i,j), r(i,j));
+      }
+}
+
+
+TEST(TestSTensorFunc, Evaluate3D)
+{
+  const char* func = "sin(x) | cos (y) | exp(z) |"
+                     "sin(x)*sin(y) | sin(x)*cos(y) | exp(x)*exp(y)";
+
+  STensorFuncExpr f(func);
+
+  for (double x : {0.1, 0.2, 0.3})
+    for (double y : {0.5, 0.6, 0.7})
+      for (double z : {0.8, 0.9, 1.0}) {
+        const Vec3 X(x,y,z);
+        const SymmTensor fx = f(X);
+        const Tensor r({sin(x), sin(x)*sin(y), exp(x)*exp(y),
+                        sin(x)*sin(y), cos(y), sin(x)*cos(y),
+                        exp(x)*exp(y), sin(x)*cos(y), exp(z)});
+        for (size_t i = 1; i <= 3; ++i)
+          for (size_t j = 1; j <= 3; ++j)
+            EXPECT_DOUBLE_EQ(fx(i,j), r(i,j));
+      }
+}
+
+
+TEST(TestSTensorFunction, Gradient2D)
+{
+  const char* g   = "sin(x)*sin(y) | exp(x)*exp(2*y) | x*x*y*y";
+  const char* g_x = "cos(x)*sin(y) | exp(x)*exp(2*y) | 2*x*y*y";
+  const char* g_y = "sin(x)*cos(y) | 2*exp(x)*exp(2*y) | 2*x*x*y";
+
+  STensorFuncExpr f(g);
+  f.addDerivative(g_x,"",1);
+  f.addDerivative(g_y,"",2);
+
+  EXPECT_TRUE(f.isConstant());
+
+  for (double x : {0.1, 0.2, 0.3})
+    for (double y : {0.5, 0.6, 0.7}) {
+      const Vec3 X(x,y);
+      utl::matrix3d<Real> r(2,2,2);
+      r.fill(std::array{cos(x)*sin(y), 2*x*y*y, 2*x*y*y, exp(x)*exp(2*y),
+                        sin(x)*cos(y), 2*x*x*y, 2*x*x*y, 2*exp(x)*exp(2*y)}.data());
+
+      for (size_t d = 1; d <= 2; ++d) {
+        const SymmTensor dx = f.deriv(X,d);
+        for (size_t i = 1; i <= 2; ++i)
+          for (size_t j = 1; j <= 2; ++j)
+            EXPECT_DOUBLE_EQ(dx(i,j), r(i,j,d));
+      }
+    }
+}
+
+
+TEST(TestSTensorFunction, Gradient2DFD)
+{
+  const char* g   = "sin(x)*sin(y) | exp(x)*exp(2*y) | x*x*y*y";
+
+  const double eps = 1e-6;
+  STensorFuncExpr f(g,"",eps);
+
+  EXPECT_TRUE(f.isConstant());
+
+  for (double x : {0.1, 0.2, 0.3})
+    for (double y : {0.5, 0.6, 0.7}) {
+      const Vec3 X(x,y);
+      const Vec3 Xp(x + 0.5*eps, y + 0.5*eps);
+      const Vec3 Xm(x - 0.5*eps, y - 0.5*eps);
+      utl::matrix3d<Real> r(2,2,2);
+      r.fill(std::array{(sin(Xp.x) - sin(Xm.x))*sin(y),
+                        (Xp.x*Xp.x - Xm.x*Xm.x)*y*y,
+                        (Xp.x*Xp.x - Xm.x*Xm.x)*y*y,
+                        (exp(Xp.x) - exp(Xm.x))*exp(2*y),
+                        sin(x)*(sin(Xp.y) - sin(Xm.y)),
+                        x*x*(Xp.y*Xp.y - Xm.y*Xm.y),
+                        x*x*(Xp.y*Xp.y - Xm.y*Xm.y),
+                        exp(x)*(exp(2*Xp.y) - exp(2*Xm.y))}.data());
+      r.multiply(1.0 / eps);
+
+      for (size_t d = 1; d <= 2; ++d) {
+        const SymmTensor dx = f.deriv(X,d);
+        for (size_t i = 1; i <= 2; ++i)
+          for (size_t j = 1; j <= 2; ++j)
+            EXPECT_NEAR(dx(i,j), r(i,j,d), 1e-8);
+      }
+    }
+}
+
```

Comment:
Add extra blank line here, for consistency

Reply:
fixed.